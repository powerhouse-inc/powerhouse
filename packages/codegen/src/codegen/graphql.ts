import { type CodegenConfig, generate } from "@graphql-codegen/cli";
import { TypeScriptPluginConfig } from "@graphql-codegen/typescript";
import { readdirSync } from "fs";

const getDirectories = (source: string) =>
  readdirSync(source, { withFileTypes: true })
    .filter((dirent) => dirent.isDirectory())
    .map((dirent) => dirent.name);

const tsConfig: TypeScriptPluginConfig = {
  strictScalars: true,
  scalars: {
    Unknown: "unknown",
    DateTime: "string",
    Attachment: "string",
    Address: "`${string}:0x${string}`",
  },
  enumsAsTypes: true,
  allowEnumStringTypes: true,
  avoidOptionals: {
    field: true,
  },
  skipTypename: true,
  // maybeValue: "T | null | undefined",
  inputMaybeValue: "T | null | undefined",
};

function schemaConfig(name: string, dir: string): CodegenConfig["generates"] {
  return {
    [`${dir}/${name}/gen/schema/types.ts`]: {
      schema: [
        {
          [`${dir}/${name}/schema.graphql`]: {
            skipGraphQLImport: false,
          },
        },
      ],
      plugins: ["typescript"],
      config: tsConfig,
    },
    [`${dir}/${name}/gen/schema/zod.ts`]: {
      schema: `${dir}/${name}/schema.graphql`,
      plugins: ["@acaldas/graphql-codegen-typescript-validation-schema"],
      config: {
        importFrom: `./types`,
        schema: "zod",
        ...tsConfig,
        scalarSchemas: {
          Unknown: "z.unknown()",
          DateTime: "z.string().datetime()",
          Attachment: "z.string()",
          Address:
            "z.custom<`${string}:0x${string}`>((val) => /^[a-zA-Z0-9]+:0x[a-fA-F0-9]{40}$/.test(val as string))",
        },
        directives: {
          equals: {
            value: ["regex", "/^$1$/"],
          },
        },
        withObjectType: true,
      },
    },
  };
}

export const generateSchema = (
  model: string,
  dir: string,
  { watch = false, format = false } = {},
) => {
  const documentModelConfig = schemaConfig(model, dir);
  const config: CodegenConfig = {
    overwrite: true,
    generates: documentModelConfig,
    watch,
    hooks: {
      afterOneFileWrite: format ? ["prettier --ignore-path --write"] : [],
    },
  };
  return generate(config, true);
};

export const generateSchemas = (
  dir: string,
  { watch = false, format = false } = {},
) => {
  const documentModels = getDirectories(dir);
  const documentModelConfigs = documentModels.reduce(
    (obj, model) => ({
      ...obj,
      ...schemaConfig(model, dir),
    }),
    {},
  );

  const config: CodegenConfig = {
    overwrite: true,
    generates: documentModelConfigs,
    watch,
    hooks: {
      afterOneFileWrite: format ? ["prettier --ignore-path --write"] : [],
    },
  };
  return generate(config, true);
};
