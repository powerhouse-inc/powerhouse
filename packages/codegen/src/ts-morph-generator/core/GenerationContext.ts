import {
  type DocumentModelState,
  type Module,
  type OperationError,
} from "document-model";
import { type Project } from "ts-morph";

export type CodegenOperation = {
  id: string;
  name: string | null;
  description: string | null;
  examples: any[];
  reducer: any;
  schema: string | null;
  template: string | null;
  hasInput: boolean;
  hasAttachment: boolean | undefined;
  scope: string;
  state: string;
  errors?: OperationError[];
};

export interface GenerationContext {
  rootDir: string;
  docModel: DocumentModelState;
  module: Module;
  project: Project;
  operations: CodegenOperation[];
  forceUpdate: boolean;
}

export type PHProjectDirectories = {
  documentModelDir?: string;
  editorsDir?: string;
  processorsDir?: string;
  subgraphsDir?: string;
};

export type CodeGeneratorOptions = {
  directories?: PHProjectDirectories;
  forceUpdate?: boolean;
};
