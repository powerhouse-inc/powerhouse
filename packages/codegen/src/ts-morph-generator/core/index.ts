export * from "./FileGenerator.js";
export * from "./GenerationContext.js";
export * from "./TSMorphCodeGenerator.js";

