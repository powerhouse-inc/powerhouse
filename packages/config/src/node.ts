import { readFileSync, writeFileSync } from "node:fs";
import type { PowerhouseConfig } from "./powerhouse.js";
import { DEFAULT_CONFIG } from "./powerhouse.js";

export function getConfig(path = "./powerhouse.config.json") {
  let config: PowerhouseConfig = { ...DEFAULT_CONFIG };
  try {
    const configStr = readFileSync(path, "utf-8");
    const userConfig = JSON.parse(configStr) as PowerhouseConfig;
    config = { ...config, ...userConfig };
  } catch {
    console.warn("No powerhouse.config.json found, using defaults");
  }

  return config;
}

export function writeConfig(
  config: PowerhouseConfig,
  path = "./powerhouse.config.json",
) {
  writeFileSync(path, JSON.stringify(config, null, 4));
}
