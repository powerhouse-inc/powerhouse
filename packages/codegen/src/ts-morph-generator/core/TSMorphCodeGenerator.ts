import { DirectoryManager, ImportManager } from "@powerhousedao/codegen";
import type {
  DocumentModelGlobalState,
  ModuleSpecification,
} from "document-model";
import fs from "fs/promises";
import { Project } from "ts-morph";
import type { FileGenerator } from "./FileGenerator.js";
import type {
  CodeGeneratorOptions,
  CodegenOperation,
  GenerationContext,
  PHProjectDirectories,
} from "./GenerationContext.js";
import { ReducerGenerator } from "./ReducerGenerator.js";

export class TSMorphCodeGenerator {
  private project = new Project();
  private generators = new Map<string, FileGenerator>();
  private directories: PHProjectDirectories = {
    documentModelDir: "document-model",
    editorsDir: "editors",
    processorsDir: "processors",
    subgraphsDir: "subgraphs",
  };
  private forceUpdate = false;

  constructor(
    private rootDir: string,
    private docModels: DocumentModelGlobalState[],
    options: CodeGeneratorOptions = { directories: {}, forceUpdate: false },
  ) {
    this.directories = {
      ...this.directories,
      ...options.directories,
    };

    this.forceUpdate = options.forceUpdate ?? false;

    this.setupGenerators();
  }

  private setupGenerators(): void {
    const importManager = new ImportManager();
    const directoryManager = new DirectoryManager(this.directories);

    // Register all generators
    this.generators.set(
      "reducers",
      new ReducerGenerator(importManager, directoryManager),
    );
  }

  // Generate specific file types
  async generateReducers(): Promise<void> {
    await this.generateFileType("reducers");
  }

  // Generate everything
  async generateAll(): Promise<void> {
    for (const [type] of this.generators) {
      await this.generateFileType(type);
    }
  }

  private async generateFileType(type: string): Promise<void> {
    const generator = this.generators.get(type);
    if (!generator) {
      throw new Error(`No generator registered for type: ${type}`);
    }

    await this.setupProject();

    for (const docModel of this.docModels) {
      const latestSpec =
        docModel.specifications[docModel.specifications.length - 1];

      for (const module of latestSpec.modules) {
        const context = this.createGenerationContext(
          docModel,
          module,
          this.forceUpdate,
        );

        await generator.generate(context);
      }
    }
  }

  private async setupProject(): Promise<void> {
    // Only load files from configured directories
    const sourcePaths: string[] = [];

    if (this.directories.documentModelDir) {
      const dirPath = `${this.rootDir}/${this.directories.documentModelDir}`;
      await this.ensureDirectoryExists(dirPath);
      sourcePaths.push(`${dirPath}/**/*.ts`);
    }
    if (this.directories.editorsDir) {
      const dirPath = `${this.rootDir}/${this.directories.editorsDir}`;
      await this.ensureDirectoryExists(dirPath);
      sourcePaths.push(`${dirPath}/**/*.ts`);
    }
    if (this.directories.processorsDir) {
      const dirPath = `${this.rootDir}/${this.directories.processorsDir}`;
      await this.ensureDirectoryExists(dirPath);
      sourcePaths.push(`${dirPath}/**/*.ts`);
    }
    if (this.directories.subgraphsDir) {
      const dirPath = `${this.rootDir}/${this.directories.subgraphsDir}`;
      await this.ensureDirectoryExists(dirPath);
      sourcePaths.push(`${dirPath}/**/*.ts`);
    }

    // Exclude node_modules from all paths
    sourcePaths.push(`!${this.rootDir}/**/node_modules/**`);

    if (sourcePaths.length > 0) {
      this.project.addSourceFilesAtPaths(sourcePaths);
    }
  }

  private createGenerationContext(
    docModel: DocumentModelGlobalState,
    module: ModuleSpecification,
    forceUpdate = false,
  ): GenerationContext {
    const operations: CodegenOperation[] = module.operations.map((op) => ({
      ...op,
      hasInput: op.schema !== null,
      hasAttachment: op.schema?.includes(": Attachment"),
      scope: op.scope || "global",
      state: op.scope === "global" ? "" : op.scope,
    }));

    return {
      rootDir: this.rootDir,
      docModel,
      module,
      project: this.project,
      operations,
      forceUpdate,
    };
  }

  private async ensureDirectoryExists(dirPath: string): Promise<void> {
    try {
      await fs.mkdir(dirPath, { recursive: true });
    } catch (err) {
      console.error(`Failed to create directory: ${dirPath}`, err);
      throw err;
    }
  }
}
