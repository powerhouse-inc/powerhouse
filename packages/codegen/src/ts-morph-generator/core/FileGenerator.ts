import { type DirectoryManager } from "../utilities/DirectoryManager.js";
import { type ImportManager } from "../utilities/ImportManager.js";
import { type GenerationContext } from "./GenerationContext.js";

export abstract class FileGenerator {
  constructor(
    protected importManager: ImportManager,
    protected directoryManager: DirectoryManager,
  ) {}

  abstract generate(context: GenerationContext): Promise<void>;
}
