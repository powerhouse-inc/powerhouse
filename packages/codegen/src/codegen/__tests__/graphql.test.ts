import * as validationPlugin from "@acaldas/graphql-codegen-typescript-validation-schema";
import { codegen } from "@graphql-codegen/core";
import * as typescriptPlugin from "@graphql-codegen/typescript";
import { parse } from "graphql";
import { readFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import { scalars, tsConfig, zodConfig } from "../graphql.js";

const documentDriveSchema = readFileSync(
  path.join(fileURLToPath(import.meta.url), "../schema.graphql"),
  "utf-8",
);

describe("graphql codegen", () => {
  it("should generate correct typescript code for schema", async () => {
    const schema = `
      ${Object.keys(scalars)
        .map((scalar) => `scalar ${scalar}`)
        .join("\n")}
      type User {
        id: ID!
        name: String!
        email: String
        address: Address
        createdAt: DateTime!
      }
      
      type Query {
        user(id: ID!): User
        users: [User!]!
      }
    `;

    const result = await codegen({
      filename: "",
      plugins: [
        {
          typescript: tsConfig,
        },
      ],
      schema: parse(schema),
      documents: [],
      config: {},
      pluginMap: {
        typescript: typescriptPlugin,
      },
    });
    expect(result).toMatchSnapshot();
  });

  it("should generate correct zod code for schema", async () => {
    const schema = `
      ${Object.keys(scalars)
        .map((scalar) => `scalar ${scalar}`)
        .join("\n")}
      type User {
        id: ID!
        name: String!
        email: String
        address: Address
        createdAt: DateTime!
      }
      
      type Query {
        user(id: ID!): User
        users: [User!]!
      }
    `;

    const result = await codegen({
      filename: "",
      plugins: [
        {
          "graphql-codegen-typescript-validation-schema": zodConfig,
        },
      ],
      schema: parse(schema),
      documents: [],
      config: {},
      pluginMap: {
        "graphql-codegen-typescript-validation-schema": validationPlugin,
      },
    });
    expect(result).toMatchSnapshot();
  });

  it("should generate correct typescript code for drive schema", async () => {
    const result = await codegen({
      filename: "",
      plugins: [
        {
          typescript: tsConfig,
        },
      ],
      schema: parse(documentDriveSchema),
      documents: [],
      config: {},
      pluginMap: {
        typescript: typescriptPlugin,
      },
    });
    expect(result).toMatchSnapshot();
  });
  it("should generate correct zod code for drive schema", async () => {
    const result = await codegen({
      filename: "",
      plugins: [
        {
          "@acaldas/graphql-codegen-typescript-validation-schema": zodConfig,
        },
      ],
      schema: parse(documentDriveSchema),
      documents: [],
      config: {},
      pluginMap: {
        "@acaldas/graphql-codegen-typescript-validation-schema":
          validationPlugin,
      },
    });
    expect(result).toMatchSnapshot();
  });
});
