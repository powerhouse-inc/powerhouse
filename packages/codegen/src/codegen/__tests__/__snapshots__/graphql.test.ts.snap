// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`graphql codegen > should generate correct typescript code for drive schema 1`] = `
"export type Maybe<T> = T | null;
export type InputMaybe<T> = T | null | undefined;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type MakeEmpty<T extends { [key: string]: unknown }, K extends keyof T> = { [_ in K]?: never };
export type Incremental<T> = T | { [P in keyof T]?: P extends ' $fragmentName' | '__typename' ? T[P] : never };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: { input: string; output: string; }
  String: { input: string; output: string; }
  Boolean: { input: boolean; output: boolean; }
  Int: { input: number; output: number; }
  Float: { input: number; output: number; }
  Unknown: { input: unknown; output: unknown; }
};

export type FolderNode = {
  id: Scalars['String']['output'];
  name: Scalars['String']['output'];
  kind: Scalars['String']['output'];
  parentFolder: Maybe<Scalars['String']['output']>;
};

export type SynchronizationUnit = {
  syncId: Scalars['ID']['output'];
  scope: Scalars['String']['output'];
  branch: Scalars['String']['output'];
};

export type FileNode = {
  id: Scalars['String']['output'];
  name: Scalars['String']['output'];
  kind: Scalars['String']['output'];
  documentType: Scalars['String']['output'];
  parentFolder: Maybe<Scalars['String']['output']>;
  synchronizationUnits: Array<SynchronizationUnit>;
};

export type Node = FolderNode | FileNode;

export type DocumentDriveState = {
  id: Scalars['ID']['output'];
  name: Scalars['String']['output'];
  nodes: Array<Node>;
  icon: Maybe<Scalars['String']['output']>;
};

export type ListenerFilter = {
  documentType: Array<Scalars['String']['output']>;
  documentId: Maybe<Array<Scalars['ID']['output']>>;
  scope: Maybe<Array<Scalars['String']['output']>>;
  branch: Maybe<Array<Scalars['String']['output']>>;
};

export type TransmitterType =
  | 'Internal'
  | 'SwitchboardPush'
  | 'PullResponder'
  | 'SecureConnect'
  | 'MatrixConnect'
  | 'RESTWebhook';

export type ListenerCallInfo = {
  transmitterType: Maybe<TransmitterType | \`\${TransmitterType}\`>;
  name: Maybe<Scalars['String']['output']>;
  data: Maybe<Scalars['String']['output']>;
};

export type Listener = {
  listenerId: Scalars['ID']['output'];
  label: Maybe<Scalars['String']['output']>;
  block: Scalars['Boolean']['output'];
  system: Scalars['Boolean']['output'];
  filter: ListenerFilter;
  callInfo: Maybe<ListenerCallInfo>;
};

export type TriggerType =
  | 'PullResponder';

export type PullResponderTriggerData = {
  listenerId: Scalars['ID']['output'];
  url: Scalars['String']['output'];
  interval: Scalars['String']['output'];
};

export type TriggerData = PullResponderTriggerData;

export type Trigger = {
  id: Scalars['ID']['output'];
  type: TriggerType | \`\${TriggerType}\`;
  data: Maybe<TriggerData>;
};

export type DocumentDriveLocalState = {
  sharingType: Maybe<Scalars['String']['output']>;
  availableOffline: Scalars['Boolean']['output'];
  listeners: Array<Listener>;
  triggers: Array<Trigger>;
};

export type AddFileInput = {
  id: Scalars['ID']['input'];
  name: Scalars['String']['input'];
  documentType: Scalars['String']['input'];
  parentFolder?: InputMaybe<Scalars['ID']['input']>;
  synchronizationUnits: Array<SynchronizationUnit>;
  document?: InputMaybe<Scalars['Unknown']['input']>;
};

export type AddFolderInput = {
  id: Scalars['ID']['input'];
  name: Scalars['String']['input'];
  parentFolder?: InputMaybe<Scalars['ID']['input']>;
};

export type DeleteNodeInput = {
  id: Scalars['ID']['input'];
};

export type UpdateFileInput = {
  id: Scalars['ID']['input'];
  parentFolder?: InputMaybe<Scalars['ID']['input']>;
  name?: InputMaybe<Scalars['String']['input']>;
  documentType?: InputMaybe<Scalars['String']['input']>;
};

export type UpdateNodeInput = {
  id: Scalars['ID']['input'];
  parentFolder?: InputMaybe<Scalars['ID']['input']>;
  name?: InputMaybe<Scalars['String']['input']>;
};

export type CopyNodeInput = {
  srcId: Scalars['ID']['input'];
  targetId: Scalars['ID']['input'];
  targetName?: InputMaybe<Scalars['String']['input']>;
  targetParentFolder?: InputMaybe<Scalars['ID']['input']>;
  synchronizationUnits?: InputMaybe<Array<SynchronizationUnit>>;
};

export type MoveNodeInput = {
  srcFolder: Scalars['ID']['input'];
  targetParentFolder?: InputMaybe<Scalars['ID']['input']>;
};

export type SetDriveNameInput = {
  name: Scalars['String']['input'];
};

export type SetSharingTypeInput = {
  type: Scalars['String']['input'];
};

export type SetDriveIconInput = {
  icon: Scalars['String']['input'];
};

export type SetAvailableOfflineInput = {
  availableOffline: Scalars['Boolean']['input'];
};

export type AddListenerInput = {
  listener: Listener;
};

export type RemoveListenerInput = {
  listenerId: Scalars['String']['input'];
};

export type AddTriggerInput = {
  trigger: Trigger;
};

export type RemoveTriggerInput = {
  triggerId: Scalars['String']['input'];
};
"
`;

exports[`graphql codegen > should generate correct typescript code for schema 1`] = `
"export type Maybe<T> = T | null;
export type InputMaybe<T> = T | null | undefined;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type MakeEmpty<T extends { [key: string]: unknown }, K extends keyof T> = { [_ in K]?: never };
export type Incremental<T> = T | { [P in keyof T]?: P extends ' $fragmentName' | '__typename' ? T[P] : never };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: { input: string; output: string; }
  String: { input: string; output: string; }
  Boolean: { input: boolean; output: boolean; }
  Int: { input: number; output: number; }
  Float: { input: number; output: number; }
  Unknown: { input: unknown; output: unknown; }
  DateTime: { input: string; output: string; }
  Attachment: { input: string; output: string; }
  Address: { input: \`\${string}:0x\${string}\`; output: \`\${string}:0x\${string}\`; }
  Amount_Tokens: { input: number; output: number; }
  EthereumAddress: { input: string; output: string; }
  EmailAddress: { input: string; output: string; }
  Amount_Percentage: { input: number; output: number; }
  Date: { input: string; output: string; }
  URL: { input: string; output: string; }
  Amount_Money: { input: number; output: number; }
  OLabel: { input: string; output: string; }
  Currency: { input: string; output: string; }
  PHID: { input: string; output: string; }
  OID: { input: string; output: string; }
  Amount_Fiat: { input: { unit: string, value: number }; output: { unit: string, value: number }; }
  Amount_Currency: { input: { unit: string, value: string }; output: { unit: string, value: string }; }
  Amount_Crypto: { input: { unit: string, value: string }; output: { unit: string, value: string }; }
  Amount: { input: { unit?: string, value?: number }; output: { unit?: string, value?: number }; }
  Upload: { input: File; output: File; }
};

export type User = {
  id: Scalars['ID']['output'];
  name: Scalars['String']['output'];
  email: Maybe<Scalars['String']['output']>;
  address: Maybe<Scalars['Address']['output']>;
  createdAt: Scalars['DateTime']['output'];
};

export type Query = {
  user: Maybe<User>;
  users: Array<User>;
};


export type QueryUserArgs = {
  id: Scalars['ID']['input'];
};
"
`;

exports[`graphql codegen > should generate correct zod code for drive schema 1`] = `
"import { z } from 'zod'
import type { FolderNode, SynchronizationUnit, FileNode, DocumentDriveState, ListenerFilter, TransmitterType, ListenerCallInfo, Listener, TriggerType, PullResponderTriggerData, Trigger, DocumentDriveLocalState, AddFileInput, AddFolderInput, DeleteNodeInput, UpdateFileInput, UpdateNodeInput, CopyNodeInput, MoveNodeInput, SetDriveNameInput, SetSharingTypeInput, SetDriveIconInput, SetAvailableOfflineInput, AddListenerInput, RemoveListenerInput, AddTriggerInput, RemoveTriggerInput } from './types.js'

type Properties<T> = Required<{
  [K in keyof T]: z.ZodType<T[K], any, T[K]>;
}>;

type definedNonNullAny = {};

export const isDefinedNonNullAny = (v: any): v is definedNonNullAny => v !== undefined && v !== null;

export const definedNonNullAnySchema = z.any().refine((v) => isDefinedNonNullAny(v));

export const TransmitterTypeSchema = z.enum(['Internal', 'SwitchboardPush', 'PullResponder', 'SecureConnect', 'MatrixConnect', 'RESTWebhook']);

export const TriggerTypeSchema = z.enum(['PullResponder']);

export function FolderNodeSchema(): z.ZodObject<Properties<FolderNode>> {
  return z.object({
    __typename: z.literal('FolderNode').optional(),
    id: z.string(),
    name: z.string(),
    kind: z.string(),
    parentFolder: z.string().nullable()
  })
}

export function SynchronizationUnitSchema(): z.ZodObject<Properties<SynchronizationUnit>> {
  return z.object({
    __typename: z.literal('SynchronizationUnit').optional(),
    syncId: z.string(),
    scope: z.string(),
    branch: z.string()
  })
}

export function FileNodeSchema(): z.ZodObject<Properties<FileNode>> {
  return z.object({
    __typename: z.literal('FileNode').optional(),
    id: z.string(),
    name: z.string(),
    kind: z.string(),
    documentType: z.string(),
    parentFolder: z.string().nullable(),
    synchronizationUnits: z.array(SynchronizationUnitSchema())
  })
}

export function NodeSchema() {
  return z.union([FolderNodeSchema(), FileNodeSchema()])
}

export function DocumentDriveStateSchema(): z.ZodObject<Properties<DocumentDriveState>> {
  return z.object({
    __typename: z.literal('DocumentDriveState').optional(),
    id: z.string(),
    name: z.string(),
    nodes: z.array(NodeSchema()),
    icon: z.string().nullable()
  })
}

export function ListenerFilterSchema(): z.ZodObject<Properties<ListenerFilter>> {
  return z.object({
    __typename: z.literal('ListenerFilter').optional(),
    documentType: z.array(z.string()),
    documentId: z.array(z.string()).nullable(),
    scope: z.array(z.string()).nullable(),
    branch: z.array(z.string()).nullable()
  })
}

export function ListenerCallInfoSchema(): z.ZodObject<Properties<ListenerCallInfo>> {
  return z.object({
    __typename: z.literal('ListenerCallInfo').optional(),
    transmitterType: TransmitterTypeSchema.nullable(),
    name: z.string().nullable(),
    data: z.string().nullable()
  })
}

export function ListenerSchema(): z.ZodObject<Properties<Listener>> {
  return z.object({
    __typename: z.literal('Listener').optional(),
    listenerId: z.string(),
    label: z.string().nullable(),
    block: z.boolean(),
    system: z.boolean(),
    filter: ListenerFilterSchema(),
    callInfo: ListenerCallInfoSchema().nullable()
  })
}

export function PullResponderTriggerDataSchema(): z.ZodObject<Properties<PullResponderTriggerData>> {
  return z.object({
    __typename: z.literal('PullResponderTriggerData').optional(),
    listenerId: z.string(),
    url: z.string(),
    interval: z.string()
  })
}

export function TriggerDataSchema() {
  return PullResponderTriggerDataSchema()
}

export function TriggerSchema(): z.ZodObject<Properties<Trigger>> {
  return z.object({
    __typename: z.literal('Trigger').optional(),
    id: z.string(),
    type: TriggerTypeSchema,
    data: TriggerDataSchema().nullable()
  })
}

export function DocumentDriveLocalStateSchema(): z.ZodObject<Properties<DocumentDriveLocalState>> {
  return z.object({
    __typename: z.literal('DocumentDriveLocalState').optional(),
    sharingType: z.string().nullable(),
    availableOffline: z.boolean(),
    listeners: z.array(ListenerSchema()),
    triggers: z.array(TriggerSchema())
  })
}

export function AddFileInputSchema(): z.ZodObject<Properties<AddFileInput>> {
  return z.object({
    id: z.string(),
    name: z.string(),
    documentType: z.string(),
    parentFolder: z.string().nullish(),
    synchronizationUnits: z.array(SynchronizationUnitSchema()),
    document: z.unknown().nullish()
  })
}

export function AddFolderInputSchema(): z.ZodObject<Properties<AddFolderInput>> {
  return z.object({
    id: z.string(),
    name: z.string(),
    parentFolder: z.string().nullish()
  })
}

export function DeleteNodeInputSchema(): z.ZodObject<Properties<DeleteNodeInput>> {
  return z.object({
    id: z.string()
  })
}

export function UpdateFileInputSchema(): z.ZodObject<Properties<UpdateFileInput>> {
  return z.object({
    id: z.string(),
    parentFolder: z.string().nullish(),
    name: z.string().nullish(),
    documentType: z.string().nullish()
  })
}

export function UpdateNodeInputSchema(): z.ZodObject<Properties<UpdateNodeInput>> {
  return z.object({
    id: z.string(),
    parentFolder: z.string().nullish(),
    name: z.string().nullish()
  })
}

export function CopyNodeInputSchema(): z.ZodObject<Properties<CopyNodeInput>> {
  return z.object({
    srcId: z.string(),
    targetId: z.string(),
    targetName: z.string().nullish(),
    targetParentFolder: z.string().nullish(),
    synchronizationUnits: z.array(SynchronizationUnitSchema()).nullish()
  })
}

export function MoveNodeInputSchema(): z.ZodObject<Properties<MoveNodeInput>> {
  return z.object({
    srcFolder: z.string(),
    targetParentFolder: z.string().nullish()
  })
}

export function SetDriveNameInputSchema(): z.ZodObject<Properties<SetDriveNameInput>> {
  return z.object({
    name: z.string()
  })
}

export function SetSharingTypeInputSchema(): z.ZodObject<Properties<SetSharingTypeInput>> {
  return z.object({
    type: z.string()
  })
}

export function SetDriveIconInputSchema(): z.ZodObject<Properties<SetDriveIconInput>> {
  return z.object({
    icon: z.string()
  })
}

export function SetAvailableOfflineInputSchema(): z.ZodObject<Properties<SetAvailableOfflineInput>> {
  return z.object({
    availableOffline: z.boolean()
  })
}

export function AddListenerInputSchema(): z.ZodObject<Properties<AddListenerInput>> {
  return z.object({
    listener: ListenerSchema()
  })
}

export function RemoveListenerInputSchema(): z.ZodObject<Properties<RemoveListenerInput>> {
  return z.object({
    listenerId: z.string()
  })
}

export function AddTriggerInputSchema(): z.ZodObject<Properties<AddTriggerInput>> {
  return z.object({
    trigger: TriggerSchema()
  })
}

export function RemoveTriggerInputSchema(): z.ZodObject<Properties<RemoveTriggerInput>> {
  return z.object({
    triggerId: z.string()
  })
}
"
`;

exports[`graphql codegen > should generate correct zod code for schema 1`] = `
"import { z } from 'zod'
import type { User } from './types.js'

type Properties<T> = Required<{
  [K in keyof T]: z.ZodType<T[K], any, T[K]>;
}>;

type definedNonNullAny = {};

export const isDefinedNonNullAny = (v: any): v is definedNonNullAny => v !== undefined && v !== null;

export const definedNonNullAnySchema = z.any().refine((v) => isDefinedNonNullAny(v));

export function UserSchema(): z.ZodObject<Properties<User>> {
  return z.object({
    __typename: z.literal('User').optional(),
    id: z.string(),
    name: z.string(),
    email: z.string().nullable(),
    address: z.custom<\`\${string}:0x\${string}\`>((val) => /^[a-zA-Z0-9]+:0x[a-fA-F0-9]{40}$/.test(val as string)).nullable(),
    createdAt: z.string().datetime()
  })
}
"
`;
